Closes #1317: Add update hook to delete problematic legacy global footer menu blocks.

joeparsons opened the pull request:
## Description
Adds update hook for deleting old menu block configs that prevent Quickstart Global Footer module from being installed.

We need to discuss the fact that this probably needs to a be a 2.1.x change only but that it adds a DB update that may screw up future DB update numbers in 2.0.x.

## Related Issue
Closes #1317 

## How Has This Been Tested?
Tested locally in lando.

### Steps to reproduce problem
- Clone az_quickstart
- Checkout the latest 2.0.x release tag (`2.0.18`) with git
```
git checkout 2.0.18
```
- Perform these steps with lando (the `config-distro-update` step is crucial)
```
lando start
lando install
lando drush config-distro-update
lando db-export
```
- Checkout main branch with git
- Perform these steps with lando
```
lando rebuild
lando install
lando db-import <export-file-name>
lando drush cr -y
lando drush updatedb
lando drush config-distro-update
```
- Try installing Quickstart Global Footer module -- see errors

### Test fix
- Checkout this PR branch
- Perform these steps with lando
```
lando rebuild
lando install
lando db-import <export-file-name>
lando drush cr -y
lando drush updatedb
lando drush config-distro-update
```
- Try installing Quickstart Global Footer module -- no errors

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I've added this Pull Request to the AZ Quickstart Github project and set it to "Review in progress".
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

trackleft reviewed (dismissed):
Looks fine to me, hopefully nobody is using these blocks without the global footer.
We could consider deleting them only on install of the global footer if people are using them.

joeparsons commented:
> Looks fine to me, hopefully nobody is using these blocks without the global footer. We could consider deleting them only on install of the global footer if people are using them.

I had the same thought so I initially tried implementing this in `az_global_footer_install()` but unfortunately that hook doesn't seem to fire early enough in the install / config-import process to work (installation still fails with the same errors).

joeparsons commented:
I guess another alternative that might be worth exploring would be an event subscriber in the `az_global_footer` module that subscribes to config import events and does something about the problematic block configs then...

joeparsons commented:
Suggestion from @tadean:  Only delete menu blocks if config hash matches the hash of the default config values in `az_barrio/config/optional`

joeparsons commented:
Suggestion from @trackleft:  We may be able to use [`hook_module_preinstall()`](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Extension%21module.api.php/function/hook_module_preinstall/9.2.x) in the `az_global_footer` module.

joeparsons commented:
## 2022-02-24 Updates
- I tried using [`hook_module_preinstall()`](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Extension%21module.api.php/function/hook_module_preinstall/9.2.x) but unfortunately that doesn't work because Drupal core's 
[`ModuleInstaller::install()`](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Extension%21ModuleInstaller.php/function/ModuleInstaller%3A%3Ainstall/9.2.x) method calls [`ConfigInstaller::checkConfigurationToInstall()`](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Config%21ConfigInstaller.php/function/ConfigInstaller%3A%3AcheckConfigurationToInstall/9.2.x) before invoking `hook_module_preinstall()` and `hook_install()` implementations  and [`ConfigInstaller::checkConfigurationToInstall()`](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Config%21ConfigInstaller.php/function/ConfigInstaller%3A%3AcheckConfigurationToInstall/9.2.x) is the method that throws the [`PreExistingConfigException`](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Config%21PreExistingConfigException.php/class/PreExistingConfigException/9.2.x) if config objects already exist 😞 .
- I found checking that the `default_config_hash` of the menu blocks still matched their default state before deleting them does seem to work, so I've updated this PR to include that check.  It's worth noting that the `default_config_hash` values of those blocks have changed a few times since they were first added in [Quickstart 2.0.0-beta5](https://github.com/az-digital/az_quickstart/releases/tag/2.0.0-beta5).  This means that if someone were to try to update from an earlier version of Qucikstart 2.0.x (e.g. 2.0.7) to a Quickstart 2.1.x or 2.2.x release, this update wouldn't work and they'd still need to manually delete the footer menu block configs to be able to install the Quickstart global footer module.
- I also added a little bit more verbose messaging to the `return` lines of the update hook.

### Considerations for merging
I'd be comfortable merging this as-is now if others are also comfortable with it.  These are some things that should be considered before merging:
- This PR is really only intended for `2.1.x` (and `2.2.x`).  I do _not_ think we should backport this to `2.0.x` because the footer menu blocks are still included in `az_barrio`'s `config/install` folder in the `2.0.x` branch.  If we did want to backport this change to `2.0.x`, I think we'd need a `2.0.x` specific PR that also removes those files.
- This is blocked by #1350 and also contains the changes from #1351 so it should really be merged after that PR.  Because this branch is based on that one, this PR is currently targeting the `issue/1350` branch (to simplify review) but **should be updated to target `main` before merging**.

trackleft commented on az_quickstart.install:
```diff
@@ -163,3 +163,44 @@ function az_quickstart_update_92101() {
+      }
+    }
+
+    if ($delete_blocks) {
```
It could be confused, but don't you need to unset the ones that are different from the $footer_menu_blocks array before you delete them all?

trackleft replied:
I suppose the install will break on any existing active config. NVM

trackleft commented on az_quickstart.install:
```diff
@@ -163,3 +163,44 @@ function az_quickstart_update_92101() {
+      return t('Deleted legacy global footer menu blocks. Install Quickstart Global Footer module to re-enable footer menu blocks.');
+    }
+    else {
+      return t('Legacy global footer blocks not deleted because config differences detected. Blocks will need to be manually deleted with drush in order to install Quickstart Global Footer module.');
```
We should add the drush command to the readme in case someone needs it along with some explanation about when, if ever, they need to run it.

trackleft commented:
<img width="842" alt="image" src="https://user-images.githubusercontent.com/1023167/156045471-55f62613-24c3-4979-8bf0-b7c78c4c2876.png">

joeparsons commented:
## 2022-03-02 Updates

- Adjusted update hook number to follow updated numbering convention (changed `92102` to `920102`).
- Added documentation about the known module installation issues to the README for the Quickstart global footer module.

bberndt-uaz reviewed (approved):
Tested locally as described and confirmed that the fix allows installing the Quickstart Global Footer module without errors.
